File: common/src/main/kotlin/cn/jawbreakers/candycraftce/item/JellyDungeonKeyItem.java
package cn.jawbreakers.candycraftce.item;

import com.valentin4311.candycraftmod.block.DungeonTeleporterBlock;
import com.valentin4311.candycraftmod.block.DungeonTeleporterBlock.DungeonKind;
import com.valentin4311.candycraftmod.block.DungeonTeleporterBlock.PortalRole;
import com.valentin4311.candycraftmod.world.CCDimensions;
import com.valentin4311.candycraftmod.world.DungeonProgressData;
import com.valentin4311.candycraftmod.world.DungeonProgressData.Instance;
import com.valentin4311.candycraftmod.world.DungeonProgressData.PortalRecord;
import net.minecraft.ChatFormatting;
import net.minecraft.core.BlockPos;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.network.chat.Component;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.sounds.SoundEvents;
import net.minecraft.sounds.SoundSource;
import net.minecraft.world.InteractionResult;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.item.Item;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.TooltipFlag;
import net.minecraft.world.item.context.UseOnContext;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.block.Block;
import net.minecraft.world.level.block.state.BlockState;

import javax.annotation.Nullable;
import java.util.List;

public class JellyDungeonKeyItem extends Item {
	private static final String OWNER_TAG = "CandyCraftDungeonKeyOwner";
	private static final String INSTANCE_TAG = "CandyCraftDungeonKeyInstance";
	private static final String COMPLETIONS_TAG = "CandyCraftDungeonKeyCompletions";
	private static final String ACTIVE_TAG = "CandyCraftDungeonKeyActive";
	private static final String EXHAUSTED_TAG = "CandyCraftDungeonKeyExhausted";
	private static final String PORTAL_PLACED_TAG = "CandyCraftDungeonKeyPortalPlaced";
	private final DungeonKind kind;

	public JellyDungeonKeyItem(Properties properties) {
		this(properties, false);
	}

	public JellyDungeonKeyItem(Properties properties, boolean suguardDungeon) {
		super(properties);
		this.kind = suguardDungeon ? DungeonKind.SUGUARD : DungeonKind.JELLY;
	}

	@Override
	public InteractionResult useOn(UseOnContext context) {
		Level level = context.getLevel();
		if (context.getClickedFace() != net.minecraft.core.Direction.UP || context.getPlayer() == null) {
			return InteractionResult.FAIL;
		}
		BlockPos supportPos = context.getClickedPos();
		if (!level.getBlockState(supportPos).isCollisionShapeFullBlock(level, supportPos)) {
			return InteractionResult.FAIL;
		}
		BlockPos pos = supportPos.above();
		BlockState replaced = level.getBlockState(pos);
		if (!replaced.canBeReplaced()) {
			return InteractionResult.FAIL;
		}
		if (level.isClientSide) {
			return InteractionResult.SUCCESS;
		}

		if (!(context.getPlayer() instanceof ServerPlayer player) || !(level instanceof ServerLevel serverLevel)) {
			return InteractionResult.FAIL;
		}
		ItemStack stack = context.getItemInHand();
		refreshState(stack, player);
		CompoundTag tag = stack.getOrCreateTag();
		if (tag.getBoolean(EXHAUSTED_TAG)) {
			player.displayClientMessage(Component.translatable("message.candycraftmod.dungeon.key_exhausted"), true);
			return InteractionResult.FAIL;
		}

		ServerLevel dungeonLevel = player.server.getLevel(
				kind == DungeonKind.JELLY ? CCDimensions.JELLY_DUNGEON : CCDimensions.SUGUARD_DUNGEON);
		if (dungeonLevel == null) {
			return InteractionResult.FAIL;
		}
		DungeonProgressData data = DungeonProgressData.get(player.server);
		Instance active = data.getActive(player.getUUID(), kind);
		boolean bound = tag.contains(INSTANCE_TAG);
		if (bound && (tag.hasUUID(OWNER_TAG) && !tag.getUUID(OWNER_TAG).equals(player.getUUID())
				|| active == null || active.id() != tag.getLong(INSTANCE_TAG))) {
			tag.putBoolean(EXHAUSTED_TAG, true);
			player.displayClientMessage(Component.translatable("message.candycraftmod.dungeon.key_exhausted"), true);
			return InteractionResult.FAIL;
		}
		if (active != null && data.hasLiveEntrancePortal(player.server, player.getUUID(), kind, active.id())) {
			tag.putBoolean(PORTAL_PLACED_TAG, true);
			player.displayClientMessage(Component.translatable("message.candycraftmod.dungeon.portal_already_open"), true);
			return InteractionResult.FAIL;
		}

		Instance instance = active == null ? data.getOrCreate(player, kind) : active;
		if (!level.setBlock(pos, DungeonTeleporterBlock.state(kind, PortalRole.ENTRY), Block.UPDATE_ALL)) {
			return InteractionResult.FAIL;
		}
		bind(stack, player, instance, data.getCompletionCount(player.getUUID(), kind), true);
		data.registerPortal(serverLevel, pos, player.getUUID(), kind, instance.id());
		level.playSound(null, pos, SoundEvents.PORTAL_TRIGGER, SoundSource.BLOCKS, 1.0F, 1.25F);
		player.displayClientMessage(Component.translatable("message.candycraftmod.dungeon.portal_opened"), true);
		return InteractionResult.CONSUME;
	}

	public boolean recoverPortal(ItemStack stack, ServerPlayer player, PortalRecord portal) {
		if (portal.kind() != kind || !portal.owner().equals(player.getUUID())) {
			return false;
		}
		DungeonProgressData data = DungeonProgressData.get(player.server);
		Instance active = data.getActive(player.getUUID(), kind);
		if (active == null || active.id() != portal.instanceId()) {
			return false;
		}
		bind(stack, player, active, data.getCompletionCount(player.getUUID(), kind), false);
		return true;
	}

	public boolean matchesDungeon(DungeonKind dungeonKind) {
		return kind == dungeonKind;
	}

	@Override
	public void inventoryTick(ItemStack stack, Level level, Entity entity, int slot, boolean selected) {
		super.inventoryTick(stack, level, entity, slot, selected);
		if (!level.isClientSide && entity instanceof ServerPlayer player && player.tickCount % 10 == 0) {
			refreshState(stack, player);
		}
	}

	private void refreshState(ItemStack stack, ServerPlayer player) {
		DungeonProgressData data = DungeonProgressData.get(player.server);
		CompoundTag tag = stack.getOrCreateTag();
		Instance active = data.getActive(player.getUUID(), kind);
		boolean bound = tag.contains(INSTANCE_TAG);
		boolean exhausted = bound && (active == null || active.id() != tag.getLong(INSTANCE_TAG));
		boolean portalPlaced = bound && !exhausted
				&& data.hasLiveEntrancePortal(player.server, player.getUUID(), kind, tag.getLong(INSTANCE_TAG));
		tag.putInt(COMPLETIONS_TAG, data.getCompletionCount(player.getUUID(), kind));
		tag.putBoolean(ACTIVE_TAG, bound && !exhausted);
		tag.putBoolean(EXHAUSTED_TAG, exhausted);
		tag.putBoolean(PORTAL_PLACED_TAG, portalPlaced);
	}

	private static void bind(ItemStack stack, ServerPlayer player, Instance instance, int completions,
	                         boolean portalPlaced) {
		CompoundTag tag = stack.getOrCreateTag();
		tag.putUUID(OWNER_TAG, player.getUUID());
		tag.putLong(INSTANCE_TAG, instance.id());
		tag.putInt(COMPLETIONS_TAG, completions);
		tag.putBoolean(ACTIVE_TAG, true);
		tag.putBoolean(EXHAUSTED_TAG, false);
		tag.putBoolean(PORTAL_PLACED_TAG, portalPlaced);
	}

	@Override
	public boolean isFoil(ItemStack stack) {
		return false;
	}

	@Override
	public void appendHoverText(ItemStack stack, @Nullable Level level, List<Component> tooltip, TooltipFlag flag) {
		CompoundTag tag = stack.getTag();
		int completions = tag == null ? 0 : tag.getInt(COMPLETIONS_TAG);
		tooltip.add(Component.translatable(
				"tooltip.candycraftmod.dungeon_key.progress." + kind.getSerializedName(), completions
		).withStyle(ChatFormatting.GRAY));
		if (tag != null && tag.getBoolean(EXHAUSTED_TAG)) {
			tooltip.add(Component.translatable("tooltip.candycraftmod.dungeon_key.exhausted").withStyle(ChatFormatting.RED));
		} else if (tag != null && tag.getBoolean(ACTIVE_TAG) && tag.getBoolean(PORTAL_PLACED_TAG)) {
			tooltip.add(Component.translatable("tooltip.candycraftmod.dungeon_key.active").withStyle(ChatFormatting.YELLOW));
		} else if (tag != null && tag.getBoolean(ACTIVE_TAG)) {
			tooltip.add(Component.translatable("tooltip.candycraftmod.dungeon_key.charged").withStyle(ChatFormatting.GREEN));
		}
	}
}

File: common/src/main/kotlin/cn/jawbreakers/candycraftce/item/JellyDungeonKeyItem.kt
package cn.jawbreakers.candycraftce.item

import net.minecraft.world.item.Item

class JellyDungeonKeyItem(
    properties: Properties,
    suguardDungeon: Boolean = false,
) : Item(properties) {
//    private val kind: DungeonKind
//
//    init {
//        this.kind = if (suguardDungeon) DungeonKind.SUGUARD else DungeonKind.JELLY
//    }
//
//    override fun useOn(context: net.minecraft.world.item.context.UseOnContext): net.minecraft.world.InteractionResult {
//        val level = context.getLevel()
//        if (context.getClickedFace() != net.minecraft.core.Direction.UP || context.getPlayer() == null) {
//            return net.minecraft.world.InteractionResult.FAIL
//        }
//        val supportPos = context.getClickedPos()
//        if (!level.getBlockState(supportPos).isCollisionShapeFullBlock(level, supportPos)) {
//            return net.minecraft.world.InteractionResult.FAIL
//        }
//        val pos = supportPos.above()
//        val replaced = level.getBlockState(pos)
//        if (!replaced.canBeReplaced()) {
//            return net.minecraft.world.InteractionResult.FAIL
//        }
//        if (level.isClientSide) {
//            return net.minecraft.world.InteractionResult.SUCCESS
//        }
//
//        if (context.getPlayer() !is net.minecraft.server.level.ServerPlayer || level !is net.minecraft.server.level.ServerLevel) {
//            return net.minecraft.world.InteractionResult.FAIL
//        }
//        val stack = context.getItemInHand()
//        refreshState(stack, player)
//        val tag = stack.getOrCreateTag()
//        if (tag.getBoolean(JellyDungeonKeyItem.Companion.EXHAUSTED_TAG)) {
//            player.displayClientMessage(
//                net.minecraft.network.chat.Component.translatable("message.candycraftmod.dungeon.key_exhausted"),
//                true
//            )
//            return net.minecraft.world.InteractionResult.FAIL
//        }
//
//        val dungeonLevel: net.minecraft.server.level.ServerLevel? = player.server.getLevel(
//            if (kind === DungeonKind.JELLY) CCDimensions.JELLY_DUNGEON else CCDimensions.SUGUARD_DUNGEON
//        )
//        if (dungeonLevel == null) {
//            return net.minecraft.world.InteractionResult.FAIL
//        }
//        val data: DungeonProgressData = DungeonProgressData.get(player.server)
//        val active: Instance? = data.getActive(player.getUUID(), kind)
//        val bound = tag.contains(JellyDungeonKeyItem.Companion.INSTANCE_TAG)
//        if (bound && (tag.hasUUID(JellyDungeonKeyItem.Companion.OWNER_TAG) && tag.getUUID(JellyDungeonKeyItem.Companion.OWNER_TAG) != player.getUUID() || active == null || active.id() !== tag.getLong(
//                JellyDungeonKeyItem.Companion.INSTANCE_TAG
//            ))
//        ) {
//            tag.putBoolean(JellyDungeonKeyItem.Companion.EXHAUSTED_TAG, true)
//            player.displayClientMessage(
//                net.minecraft.network.chat.Component.translatable("message.candycraftmod.dungeon.key_exhausted"),
//                true
//            )
//            return net.minecraft.world.InteractionResult.FAIL
//        }
//        if (active != null && data.hasLiveEntrancePortal(player.server, player.getUUID(), kind, active.id())) {
//            tag.putBoolean(JellyDungeonKeyItem.Companion.PORTAL_PLACED_TAG, true)
//            player.displayClientMessage(
//                net.minecraft.network.chat.Component.translatable("message.candycraftmod.dungeon.portal_already_open"),
//                true
//            )
//            return net.minecraft.world.InteractionResult.FAIL
//        }
//
//        val instance: Instance = if (active == null) data.getOrCreate(player, kind) else active
//        if (!level.setBlock(
//                pos,
//                DungeonTeleporterBlock.state(kind, PortalRole.ENTRY),
//                net.minecraft.world.level.block.CBlockTags.UPDATE_ALL
//            )
//        ) {
//            return net.minecraft.world.InteractionResult.FAIL
//        }
//        JellyDungeonKeyItem.Companion.bind(
//            stack,
//            player,
//            instance,
//            data.getCompletionCount(player.getUUID(), kind),
//            true
//        )
//        data.registerPortal(level, pos, player.getUUID(), kind, instance.id())
//        level.playSound(
//            null,
//            pos,
//            net.minecraft.sounds.SoundEvents.PORTAL_TRIGGER,
//            net.minecraft.sounds.SoundSource.BLOCKS,
//            1.0f,
//            1.25f
//        )
//        player.displayClientMessage(
//            net.minecraft.network.chat.Component.translatable("message.candycraftmod.dungeon.portal_opened"),
//            true
//        )
//        return net.minecraft.world.InteractionResult.CONSUME
//    }
//
//    fun recoverPortal(
//        stack: net.minecraft.world.item.ItemStack,
//        player: net.minecraft.server.level.ServerPlayer,
//        portal: PortalRecord
//    ): kotlin.Boolean {
//        if (portal.kind() !== kind || !portal.owner().equals(player.getUUID())) {
//            return false
//        }
//        val data: DungeonProgressData = DungeonProgressData.get(player.server)
//        val active: Instance? = data.getActive(player.getUUID(), kind)
//        if (active == null || active.id() !== portal.instanceId()) {
//            return false
//        }
//        JellyDungeonKeyItem.Companion.bind(
//            stack,
//            player,
//            active,
//            data.getCompletionCount(player.getUUID(), kind),
//            false
//        )
//        return true
//    }
//
//    fun matchesDungeon(dungeonKind: DungeonKind?): kotlin.Boolean {
//        return kind === dungeonKind
//    }
//
//    override fun inventoryTick(
//        stack: net.minecraft.world.item.ItemStack,
//        level: net.minecraft.world.level.Level,
//        entity: net.minecraft.world.entity.Entity,
//        slot: kotlin.Int,
//        selected: kotlin.Boolean
//    ) {
//        super.inventoryTick(stack, level, entity, slot, selected)
//        if (!level.isClientSide && entity is net.minecraft.server.level.ServerPlayer && entity.tickCount % 10 == 0) {
//            refreshState(stack, entity)
//        }
//    }
//
//    private fun refreshState(
//        stack: net.minecraft.world.item.ItemStack,
//        player: net.minecraft.server.level.ServerPlayer
//    ) {
//        val data: DungeonProgressData = DungeonProgressData.get(player.server)
//        val tag = stack.getOrCreateTag()
//        val active: Instance? = data.getActive(player.getUUID(), kind)
//        val bound = tag.contains(JellyDungeonKeyItem.Companion.INSTANCE_TAG)
//        val exhausted =
//            bound && (active == null || active.id() !== tag.getLong(JellyDungeonKeyItem.Companion.INSTANCE_TAG))
//        val portalPlaced = bound && !exhausted && data.hasLiveEntrancePortal(
//            player.server,
//            player.getUUID(),
//            kind,
//            tag.getLong(JellyDungeonKeyItem.Companion.INSTANCE_TAG)
//        )
//        tag.putInt(JellyDungeonKeyItem.Companion.COMPLETIONS_TAG, data.getCompletionCount(player.getUUID(), kind))
//        tag.putBoolean(JellyDungeonKeyItem.Companion.ACTIVE_TAG, bound && !exhausted)
//        tag.putBoolean(JellyDungeonKeyItem.Companion.EXHAUSTED_TAG, exhausted)
//        tag.putBoolean(JellyDungeonKeyItem.Companion.PORTAL_PLACED_TAG, portalPlaced)
//    }
//
//    override fun isFoil(stack: net.minecraft.world.item.ItemStack): kotlin.Boolean {
//        return false
//    }
//
//    override fun appendHoverText(
//        stack: net.minecraft.world.item.ItemStack,
//        level: net.minecraft.world.level.Level?,
//        tooltip: kotlin.collections.MutableList<net.minecraft.network.chat.Component?>,
//        flag: net.minecraft.world.item.TooltipFlag
//    ) {
//        val tag = stack.getTag()
//        val completions = if (tag == null) 0 else tag.getInt(JellyDungeonKeyItem.Companion.COMPLETIONS_TAG)
//        tooltip.add(
//            net.minecraft.network.chat.Component.translatable(
//                "tooltip.candycraftmod.dungeon_key.progress." + kind.getSerializedName(), completions
//            ).withStyle(net.minecraft.ChatFormatting.GRAY)
//        )
//        if (tag != null && tag.getBoolean(JellyDungeonKeyItem.Companion.EXHAUSTED_TAG)) {
//            tooltip.add(
//                net.minecraft.network.chat.Component.translatable("tooltip.candycraftmod.dungeon_key.exhausted")
//                    .withStyle(net.minecraft.ChatFormatting.RED)
//            )
//        } else if (tag != null && tag.getBoolean(JellyDungeonKeyItem.Companion.ACTIVE_TAG) && tag.getBoolean(
//                JellyDungeonKeyItem.Companion.PORTAL_PLACED_TAG
//            )
//        ) {
//            tooltip.add(
//                net.minecraft.network.chat.Component.translatable("tooltip.candycraftmod.dungeon_key.active")
//                    .withStyle(net.minecraft.ChatFormatting.YELLOW)
//            )
//        } else if (tag != null && tag.getBoolean(JellyDungeonKeyItem.Companion.ACTIVE_TAG)) {
//            tooltip.add(
//                net.minecraft.network.chat.Component.translatable("tooltip.candycraftmod.dungeon_key.charged")
//                    .withStyle(net.minecraft.ChatFormatting.GREEN)
//            )
//        }
//    }
//
//    companion object {
//        private const val OWNER_TAG = "CandyCraftDungeonKeyOwner"
//        private const val INSTANCE_TAG = "CandyCraftDungeonKeyInstance"
//        private const val COMPLETIONS_TAG = "CandyCraftDungeonKeyCompletions"
//        private const val ACTIVE_TAG = "CandyCraftDungeonKeyActive"
//        private const val EXHAUSTED_TAG = "CandyCraftDungeonKeyExhausted"
//        private const val PORTAL_PLACED_TAG = "CandyCraftDungeonKeyPortalPlaced"
//        private fun bind(
//            stack: net.minecraft.world.item.ItemStack,
//            player: net.minecraft.server.level.ServerPlayer,
//            instance: Instance,
//            completions: kotlin.Int,
//            portalPlaced: kotlin.Boolean
//        ) {
//            val tag = stack.getOrCreateTag()
//            tag.putUUID(JellyDungeonKeyItem.Companion.OWNER_TAG, player.getUUID())
//            tag.putLong(JellyDungeonKeyItem.Companion.INSTANCE_TAG, instance.id())
//            tag.putInt(JellyDungeonKeyItem.Companion.COMPLETIONS_TAG, completions)
//            tag.putBoolean(JellyDungeonKeyItem.Companion.ACTIVE_TAG, true)
//            tag.putBoolean(JellyDungeonKeyItem.Companion.EXHAUSTED_TAG, false)
//            tag.putBoolean(JellyDungeonKeyItem.Companion.PORTAL_PLACED_TAG, portalPlaced)
//        }
//    }
}
